File: typescript/workflows/src/enrichment.ts
/**
 * Customer profile enrichment calculations.
 */

export interface CustomerProfile {
  customer_id: string;
  [key: string]: unknown;
}

export interface EnrichedProfile extends CustomerProfile {
  health_score: number;
  churn_risk: "LOW" | "MEDIUM" | "HIGH";
  expansion_potential: "LOW" | "MEDIUM" | "HIGH";
}

/**
 * Calculate a health score (0-100) based on customer data.
 */
export function calculateHealthScore(
  profile: Record<string, unknown>
): number {
  let score = 50; // Base score

  // Usage factor (+/- 20 points)
  const usagePct = (profile.usage_pct as number) ?? 50;
  if (usagePct >= 80) {
    score += 20;
  } else if (usagePct >= 60) {
    score += 10;
  } else if (usagePct < 30) {
    score -= 10;
  }

  // Payment factor (+/- 15 points)
  const paymentStatus = String(profile.payment_status ?? "").toLowerCase();
  if (paymentStatus === "active") {
    score += 15;
  } else if (paymentStatus === "past_due" || paymentStatus === "failed") {
    score -= 15;
  }

  // NPS factor (+/- 10 points)
  const npsScore = profile.nps_score as number | undefined;
  if (npsScore !== undefined) {
    if (npsScore >= 9) {
      score += 10;
    } else if (npsScore >= 7) {
      score += 5;
    } else if (npsScore <= 5) {
      score -= 10;
    }
  }

  // Support tickets factor (+/- 5 points)
  const openTickets = (profile.open_tickets as number) ?? 0;
  if (openTickets === 0) {
    score += 5;
  } else if (openTickets > 3) {
    score -= 5;
  }

  return Math.max(0, Math.min(100, score));
}

/**
 * Calculate churn risk level based on customer signals.
 */
export function calculateChurnRisk(
  profile: Record<string, unknown>
): "LOW" | "MEDIUM" | "HIGH" {
  let riskScore = 0;

  // Low usage = higher risk
  const usagePct = (profile.usage_pct as number) ?? 50;
  if (usagePct < 30) {
    riskScore += 3;
  } else if (usagePct < 50) {
    riskScore += 1;
  }

  // Payment issues = higher risk
  const paymentStatus = String(profile.payment_status ?? "").toLowerCase();
  if (paymentStatus === "past_due" || paymentStatus === "failed") {
    riskScore += 3;
  }

  // Low NPS = higher risk
  const npsScore = profile.nps_score as number | undefined;
  if (npsScore !== undefined && npsScore <= 6) {
    riskScore += 2;
  }

  // Many open tickets = higher risk
  const openTickets = (profile.open_tickets as number) ?? 0;
  if (openTickets > 3) {
    riskScore += 1;
  }

  if (riskScore >= 5) {
    return "HIGH";
  } else if (riskScore >= 2) {
    return "MEDIUM";
  }
  return "LOW";
}

/**
 * Calculate expansion/upsell potential based on customer signals.
 */
export function calculateExpansionPotential(
  profile: Record<string, unknown>
): "LOW" | "MEDIUM" | "HIGH" {
  let potentialScore = 0;

  // High usage = expansion potential
  const usagePct = (profile.usage_pct as number) ?? 50;
  if (usagePct >= 80) {
    potentialScore += 3;
  } else if (usagePct >= 60) {
    potentialScore += 1;
  }

  // Lower tier plans = upgrade potential
  const plan = String(profile.plan ?? "").toLowerCase();
  if (["starter", "free", "basic"].includes(plan)) {
    potentialScore += 2;
  } else if (["business", "professional"].includes(plan)) {
    potentialScore += 1;
  }

  // High NPS = likely to expand
  const npsScore = profile.nps_score as number | undefined;
  if (npsScore !== undefined && npsScore >= 9) {
    potentialScore += 2;
  }

  // Deal stage consideration
  const dealStage = String(profile.deal_stage ?? "").toLowerCase();
  if (dealStage === "enterprise") {
    potentialScore += 1;
  }

  if (potentialScore >= 5) {
    return "HIGH";
  } else if (potentialScore >= 2) {
    return "MEDIUM";
  }
  return "LOW";
}

/**
 * Add calculated fields to a customer profile.
 */
export function enrichProfile(
  profile: Record<string, unknown>
): EnrichedProfile {
  return {
    ...profile,
    customer_id: String(profile.customer_id),
    health_score: calculateHealthScore(profile),
    churn_risk: calculateChurnRisk(profile),
    expansion_potential: calculateExpansionPotential(profile),
  };
}

File: python/workflows/enrichment.py
"""Customer profile enrichment calculations."""

from typing import Any


def calculate_health_score(profile: dict[str, Any]) -> int:
    """
    Calculate a health score (0-100) based on customer data.
    
    Factors:
    - Usage percentage (product engagement)
    - Payment status (billing health)
    - NPS score (customer satisfaction)
    - Support ticket volume (potential issues)
    """
    score = 50  # Base score
    
    # Usage factor (+/- 20 points)
    usage_pct = profile.get("usage_pct", 50)
    if usage_pct >= 80:
        score += 20
    elif usage_pct >= 60:
        score += 10
    elif usage_pct < 30:
        score -= 10
    
    # Payment factor (+/- 15 points)
    payment_status = profile.get("payment_status", "").lower()
    if payment_status == "active":
        score += 15
    elif payment_status in ("past_due", "failed"):
        score -= 15
    
    # NPS factor (+/- 10 points)
    nps_score = profile.get("nps_score")
    if nps_score is not None:
        if nps_score >= 9:
            score += 10
        elif nps_score >= 7:
            score += 5
        elif nps_score <= 5:
            score -= 10
    
    # Support tickets factor (+/- 5 points)
    open_tickets = profile.get("open_tickets", 0)
    if open_tickets == 0:
        score += 5
    elif open_tickets > 3:
        score -= 5
    
    return max(0, min(100, score))


def calculate_churn_risk(profile: dict[str, Any]) -> str:
    """
    Calculate churn risk level based on customer signals.
    
    Returns: "LOW", "MEDIUM", or "HIGH"
    """
    risk_score = 0
    
    # Low usage = higher risk
    usage_pct = profile.get("usage_pct", 50)
    if usage_pct < 30:
        risk_score += 3
    elif usage_pct < 50:
        risk_score += 1
    
    # Payment issues = higher risk
    payment_status = profile.get("payment_status", "").lower()
    if payment_status in ("past_due", "failed"):
        risk_score += 3
    
    # Low NPS = higher risk
    nps_score = profile.get("nps_score")
    if nps_score is not None and nps_score <= 6:
        risk_score += 2
    
    # Many open tickets = higher risk
    open_tickets = profile.get("open_tickets", 0)
    if open_tickets > 3:
        risk_score += 1
    
    # No recent activity = higher risk
    # (would check last_active date in real implementation)
    
    if risk_score >= 5:
        return "HIGH"
    elif risk_score >= 2:
        return "MEDIUM"
    return "LOW"


def calculate_expansion_potential(profile: dict[str, Any]) -> str:
    """
    Calculate expansion/upsell potential based on customer signals.
    
    Returns: "LOW", "MEDIUM", or "HIGH"
    """
    potential_score = 0
    
    # High usage = expansion potential
    usage_pct = profile.get("usage_pct", 50)
    if usage_pct >= 80:
        potential_score += 3
    elif usage_pct >= 60:
        potential_score += 1
    
    # Lower tier plans = upgrade potential
    plan = profile.get("plan", "").lower()
    if plan in ("starter", "free", "basic"):
        potential_score += 2
    elif plan in ("business", "professional"):
        potential_score += 1
    
    # High NPS = likely to expand
    nps_score = profile.get("nps_score")
    if nps_score is not None and nps_score >= 9:
        potential_score += 2
    
    # Deal stage consideration
    deal_stage = profile.get("deal_stage", "").lower()
    if deal_stage == "enterprise":
        potential_score += 1
    
    if potential_score >= 5:
        return "HIGH"
    elif potential_score >= 2:
        return "MEDIUM"
    return "LOW"


def enrich_profile(profile: dict[str, Any]) -> dict[str, Any]:
    """
    Add calculated fields to a customer profile.
    
    Args:
        profile: Merged customer data from all sources
        
    Returns:
        Profile with health_score, churn_risk, expansion_potential added
    """
    enriched = profile.copy()
    enriched["health_score"] = calculate_health_score(profile)
    enriched["churn_risk"] = calculate_churn_risk(profile)
    enriched["expansion_potential"] = calculate_expansion_potential(profile)
    return enriched
